File: services/credit-service/src/main/java/com/bankapp/creditservice/config/DemoDataSeeder.java
package com.bankapp.creditservice.config;

import com.bankapp.creditservice.domain.CreditApplication;
import com.bankapp.creditservice.domain.CreditStatus;
import com.bankapp.creditservice.repository.CreditApplicationRepository;
import java.math.BigDecimal;
import java.time.Instant;
import java.util.UUID;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.boot.ApplicationArguments;
import org.springframework.boot.ApplicationRunner;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

/**
 * Seeds one approved credit application when the table is empty so live/demo
 * environments show persisted scoring history immediately.
 */
@Component
public class DemoDataSeeder implements ApplicationRunner {

    private static final Logger log = LoggerFactory.getLogger(DemoDataSeeder.class);

    private final CreditApplicationRepository repository;

    public DemoDataSeeder(CreditApplicationRepository repository) {
        this.repository = repository;
    }

    @Override
    @Transactional
    public void run(ApplicationArguments args) {
        if (repository.count() > 0) {
            return;
        }

        CreditApplication demo = CreditApplication.builder()
                .id(UUID.fromString("11111111-1111-1111-1111-111111111111"))
                .requestedAmount(new BigDecimal("10000.00"))
                .monthlyIncome(new BigDecimal("3500.00"))
                .termMonths(24)
                .aiCollateralValueEur(new BigDecimal("150.00"))
                .monthlyPayment(new BigDecimal("455.23"))
                .interestRate(new BigDecimal("8.50"))
                .approvedLimit(new BigDecimal("10105.00"))
                .status(CreditStatus.APPROVED)
                .createdAt(Instant.parse("2026-01-15T10:00:00Z"))
                .build();

        repository.save(demo);
        log.info("Seeded demo credit application id={}", demo.getId());
    }
}

File: services/card-service/src/main/kotlin/com/bankapp/cardservice/config/DemoDataSeeder.kt
package com.bankapp.cardservice.config

import com.bankapp.cardservice.domain.Card
import com.bankapp.cardservice.domain.CardStatus
import com.bankapp.cardservice.repository.CardRepository
import org.slf4j.LoggerFactory
import org.springframework.boot.ApplicationArguments
import org.springframework.boot.ApplicationRunner
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.math.BigDecimal

/**
 * Seeds a small demo portfolio when the cards table is empty so the dashboard
 * is not blank on first visit (local Docker / live deploy).
 */
@Component
class DemoDataSeeder(
    private val cardRepository: CardRepository,
) : ApplicationRunner {

    private val log = LoggerFactory.getLogger(javaClass)

    @Transactional
    override fun run(args: ApplicationArguments) {
        if (cardRepository.count() > 0L) {
            return
        }

        val seeds = listOf(
            Card(
                cardNumberMasked = "**** **** **** 4242",
                balance = BigDecimal("1250.00"),
                creditLimit = BigDecimal("5000.00"),
                activeDebt = BigDecimal.ZERO,
                loanPrincipal = BigDecimal.ZERO,
                currency = "EUR",
                status = CardStatus.ACTIVE,
            ),
            Card(
                cardNumberMasked = "**** **** **** 1881",
                balance = BigDecimal("80.50"),
                creditLimit = BigDecimal("2500.00"),
                activeDebt = BigDecimal("420.00"),
                loanPrincipal = BigDecimal("400.00"),
                currency = "EUR",
                status = CardStatus.ACTIVE,
            ),
            Card(
                cardNumberMasked = "**** **** **** 9012",
                balance = BigDecimal("3000.00"),
                creditLimit = BigDecimal("7500.00"),
                activeDebt = BigDecimal.ZERO,
                loanPrincipal = BigDecimal.ZERO,
                currency = "EUR",
                status = CardStatus.ACTIVE,
            ),
        )

        cardRepository.saveAll(seeds)
        log.info("Seeded {} demo cards for empty database", seeds.size)
    }
}
